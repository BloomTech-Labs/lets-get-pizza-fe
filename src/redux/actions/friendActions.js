import * as types from '../types/friendTypes'
import API from '../../utils/API'

export const getFriendFavoriteShop = (id) => dispatch => {
    dispatch({type: types.GET_FRIEND_FAV_SHOP_START, payload: true})
    API.get(`/locations/${id}`)
        .then(res => {
            dispatch({type: types.GET_FRIEND_FAV_SHOP_SUCCESS, payload: res.data.location})
        })
        .catch(err => {
            dispatch({
                type: types.GET_FRIEND_FAV_SHOP_FAIL, 
                payload: {isLoading: false, error: "There was an error fetching favorite shop"}
        })
        })
}

export const getFriendData = (username) => dispatch => {
    dispatch({type: types.GET_FRIEND_START, payload: true})
    API.get(`/users/?username=${username}`)
            .then(res => {
                dispatch({type: types.GET_FRIEND_SUCCESS, payload: res.data.users[0]})
                return res.data.users[0].favorite_pizza_shop
            })
            .catch(err => {
                dispatch({
                    type: types.GET_FRIEND_FAIL, 
                    payload: {isLoading: false, error: "There was an error fetching friend data"}
            })
            })
}

// Events
export const eventsByFriend = (id) => dispatch => {
    dispatch({ type: types.FRIEND_EVENT_START, payload: true })
    API.get(`/events/users/${id}`)
        .then((res) => {
            const currentDate = new Date().toISOString();
            dispatch({
                type: types.FRIEND_EVENT_SUCCESS,
                payload: res.data
                    .sort((a, b) => new Date(a.start_time) - new Date(b.start_time))
                    .filter((date) => date.start_time > currentDate)
            });
        })
        .catch((err) => {
            dispatch({ 
                type: types.FRIEND_EVENT_FAIL, 
                payload: {isLoading: false, error: "There was an error fetching the events"} 
            })
        }); 
}

// Reviews
export const reviewsByFriend = id => dispatch => {
    dispatch({type: types.FRIEND_REVIEW_START, payload: true})
    API.get(`/reviews/users/${id}`)
        .then(res => {
            dispatch({
                type: types.FRIEND_REVIEW_SUCCESS,
                payload: res.data.sort((a, b) => b.id - a.id)
            })
        })
        .catch((err) => {
            dispatch({ 
                type: types.FRIEND_REVIEW_FAIL, 
                payload: {isLoading: false, error: 'There was an error fetching the reviews'} 
        })
        })
}

export const getAllFriends = id => dispatch => {
    dispatch({type: types.GET_ALL_FRIENDS_START, payload: true})

    API.get(`/friends/${id}`)
        .then(res => {
            dispatch({type: types.GET_ALL_FRIENDS_SUCCESS, payload: res.data})
        })
        .catch(err => {
            dispatch({
                type: types.GET_ALL_FRIENDS_FAIL, 
                payload: {error: 'There was an error fetching friends', isLoading: false}
            })
        })
}