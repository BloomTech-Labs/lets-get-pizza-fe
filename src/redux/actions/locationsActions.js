import * as types from "../types/locationTypes";
import API from "../../utils/API";

export const locationLogin = ({ username, password }, history) => (
  dispatch
) => {
  dispatch({ type: types.LOCATION_LOGIN_START, payload: true });
  API.post("auth/location/login", { username, password })
    .then((res) => {
      localStorage.setItem("token", JSON.stringify(res.data.token));
      dispatch({
        type: types.LOCATION_LOGIN_SUCCESS,
        payload: res.data.location,
      });
      window.location.replace("/");
    })
    .catch((err) => {
      dispatch({
        type: types.LOCATION_LOGIN_FAIL,
        payload: { error: "Invalid username or password", isLoading: false },
      });
    });
};

export const locationRegister = (data) => (dispatch) => {
  dispatch({ type: types.LOCATION_REGISTER_START, payload: true });
  delete data.verify_password;
  API.post("auth/location/register", data)
    .then((res) => {
      console.log(res);
      localStorage.setItem("token", JSON.stringify(res.data.token));
      dispatch({
        type: types.LOCATION_REGISTER_SUCCESS,
        payload: res.data.location,
      });
      window.location.replace("/");
    })
    .catch((err) => {
      dispatch({
        type: types.LOCATION_REGISTER_FAIL,
        payload: {
          error: "There was an error with your registration, please try again",
          isLoading: false,
        },
      });
    });
};

export const locationClaim = (data, id) => (dispatch) => {
  dispatch({ type: types.LOCATION_CLAIM_START, payload: true });
  delete data.verify_password;
  API.post(`/auth/location/claim/${id}`, data)
    .then((res) => {
      localStorage.removeItem("persist:root");
      localStorage.removeItem("token");
      localStorage.setItem("token", JSON.stringify(res.data.token));
      dispatch({
        type: types.LOCATION_CLAIM_SUCCESS,
        payload: res.data.location,
      });
      window.location.replace("/");
    })
    .catch((err) => {
      dispatch({
        type: types.LOCATION_CLAIM_FAIL,
        payload: {
          error: "There was an error claiming this business, please try again",
          isLoading: false,
        },
      });
    });
};

export const locationEvents = (id) => (dispatch) => {
  dispatch({ type: types.LOCATION_EVENTS_START, payload: true });
  API.get(`/events/locations/${id}`)
    .then((res) => {
      const currentDate = new Date().toISOString();

      dispatch({
        type: types.LOCATION_EVENTS_SUCCESS,
        payload: res.data
          .sort((a, b) => new Date(a.start_time) - new Date(b.start_time))
          .filter((date) => date.start_time > currentDate),
      });
    })
    .catch((err) => {
      dispatch({ type: types.LOCATION_EVENTS_FAIL, payload: false });
    });
};

export const locationDeleteEvent = (id, events) => (dispatch) => {
  dispatch({ type: types.LOCATION_EVENTS_DELETE_START, payload: true });
  API.delete(`/events/${id}`)
    .then((res) => {
      const filteredEvents = events.filter((event) => event.id !== id);
      dispatch({
        type: types.LOCATION_EVENTS_DELETE_SUCCESS,
        payload: filteredEvents,
      });
    })
    .catch((err) => {
      dispatch({ type: types.LOCATION_EVENTS_DELETE_FAIL, payload: false });
    });
};
