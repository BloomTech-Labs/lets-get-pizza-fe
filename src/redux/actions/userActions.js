import API from "../../utils/API";
import * as types from "../types/userTypes";

export const userLogin = ({ username, password }, history) => (dispatch) => {
  dispatch({ type: types.USER_LOGIN_START, payload: true });
  API.post("/auth/user/login", { username, password })
    .then((res) => {
      localStorage.setItem("token", JSON.stringify(res.data.token));
      dispatch({ type: types.USER_LOGIN_SUCCESS, payload: res.data.user });
      // return logged in user's id
      return res.data.user.id;
    })
    .then((id) => {
      // dispatch `getUserFriends` to grab a list of user friends
      dispatch(getUserFriends(id));
      window.location.replace("/users/dash/profile");
    })
    .catch((err) => {
      dispatch({
        type: types.USER_LOGIN_FAIL,
        payload: "Invalid username or password",
      });
    });
};

export const userRegister = (data) => (dispatch) => {
  dispatch({ type: types.USER_REGISTER_START, payload: true });
  delete data.verify_password;
  API.post("/auth/user/register", data)
    .then((res) => {
      localStorage.setItem("token", JSON.stringify(res.data.token));
      dispatch({ type: types.USER_REGISTER_SUCCESS, payload: res.data.user });
      window.location.replace("/users/dash/profile");
    })
    .catch((err) => {
      dispatch({
        type: types.USER_REGISTER_FAIL,
        payload: {
          error: "There was an error with your registration, please try again",
          isLoading: false,
        },
      });
    });
};

// Settings
export const userToggleEdit = (event, field) => (dispatch) => {
  if (event.target.id === field) {
    dispatch({ type: types.TOGGLE_EDIT, payload: "" });
  } else {
    dispatch({ type: types.TOGGLE_EDIT, payload: event.target.id });
  }
};

export const userEditSettings = (event, value) => (dispatch) => {
  !event.target.name
    ? dispatch({
        type: types.EDIT_SETTINGS,
        payload: { dietary_preference: value },
      })
    : dispatch({
        type: types.EDIT_SETTINGS,
        payload: { [event.target.name]: value },
      });
};

export const userSubmitSettings = (event, user) => (dispatch) => {
  dispatch({ type: types.SUBMIT_SETTINGS_START, payload: true });
  if (event.target.id === "save") {
    API.put(`/users`, user)
      .then((res) => {
        dispatch({ type: types.SUBMIT_SETTINGS_SUCCESS, payload: res.data });
      })
      .catch((err) => {
        dispatch({ type: types.SUBMIT_SETTINGS_FAIL, payload: false });
      });
  } else {
    dispatch({ type: types.EDIT_CANCEL_CHANGES });
  }
};

export const uploadImage = (formData, setOpen) => (dispatch) => {
  dispatch({ type: types.IMAGE_UPLOAD_START, payload: true });

  // set header `Content-Type` to `multipart/form-data`
  API.put("/users/images", formData, {
    headers: { "Content-Type": "multipart/form-data" },
  })
    .then((res) => {
      dispatch({ type: types.IMAGE_UPLOAD_SUCCESS, payload: res.data });
      setOpen(false);
    })
    .catch((err) => {
      dispatch({
        type: types.IMAGE_UPLOAD_FAIL,
        payload: {
          error: "There was an error uploading your image",
          isLoading: false,
        },
      });
    });
};

export const deleteImage = (setOpen) => (dispatch) => {
  dispatch({ type: types.IMAGE_DELETE_START, payload: true });
  API.put("/users", {
    profile_image:
      "https://res.cloudinary.com/plza/image/upload/v1588043869/qxhdqbj4sthf57bdgltz.jpg",
  })
    .then((res) => {
      dispatch({ type: types.IMAGE_DELETE_SUCCESS, payload: res.data });
      setOpen(false);
    })
    .catch((err) => {
      dispatch({
        type: types.IMAGE_DELETE_FAIL,
        payload: {
          error: "There was an error removing your image",
          isLoading: false,
        },
      });
    })
    .catch((err) => {
      dispatch({
        type: types.IMAGE_UPLOAD_FAIL,
        payload: {
          error: "There was an error uploading your image",
          isLoading: false,
        },
      });
    });
};

// Location

export const locationByUser = (id) => (dispatch) => {
  dispatch({ type: types.USER_LOCATION_START, payload: true });
  API.get(`locations/${id}`)
    .then((res) => {
      dispatch({
        type: types.USER_LOCATION_SUCCESS,
        payload: res.data.location,
      });
    })
    .catch((err) => {
      dispatch({ type: types.USER_LOCATION_FAIL, payload: false });
    });
};

// Events
export const eventsByUser = (id) => (dispatch) => {
  dispatch({ type: types.USER_EVENT_START, payload: true });
  API.get(`/events/users/${id}`)
    .then((res) => {
      const currentDate = new Date().toISOString();
      // filter out any declined invites
      const filterDeclined = res.data.invitedEvents.filter(
        (event) => event.response !== "declined"
      );

      // combine created and invited events and sort by date
      const events = [...res.data.createdEvents, ...filterDeclined]
        .sort((a, b) => new Date(a.start_time) - new Date(b.start_time))
        .filter((date) => date.start_time > currentDate);
      dispatch({
        type: types.USER_EVENT_SUCCESS,
        payload: events,
      });
    })
    .catch((err) => {
      dispatch({ type: types.USER_EVENT_FAIL, payload: false });
    });
};

export const userDeleteEvent = (id, user) => (dispatch) => {
  dispatch({ type: types.USER_EVENT_DELETE_START, payload: true });
  API.delete(`/events/${id}`)
    .then((res) => {
      const filterDeletedEvent = user.events.filter((event) => event.id !== id);
      dispatch({
        type: types.USER_EVENT_DELETE_SUCCESS,
        payload: filterDeletedEvent,
      });
    })
    .catch((err) => {
      dispatch({ type: types.USER_EVENT_DELETE_FAIL, payload: false });
    });
};

// Reviews
export const reviewsByUser = (id) => (dispatch) => {
  dispatch({type: types.USER_REVIEW_START, payload: true})
  API.get(`/reviews/users/${id}`)
    .then((res) => {
      dispatch({
        type: types.USER_REVIEW_SUCCESS,
        payload: res.data.sort((a, b) => b.id - a.id),
      });
    })
    .catch((err) => {
      dispatch({ type: types.USER_REVIEW_FAIL, payload: false });
    });
};

export const userDeleteReview = (id, user) => (dispatch) => {
  dispatch({ type: types.USER_REVIEW_DELETE_START, payload: true });
  API.delete(`/reviews/${id}`)
    .then((res) => {
      const filterDeletedReview = user.reviews.filter(
        (review) => review.id !== id
      );
      dispatch({
        type: types.USER_REVIEW_DELETE_SUCCESS,
        payload: filterDeletedReview,
      });
    })
    .catch((err) => {
      dispatch({ type: types.USER_REVIEW_DELETE_FAIL, payload: false });
    });
};

// Friends
export const getUserFriends = (id) => (dispatch) => {
  dispatch({ type: types.GET_USER_FRIENDS_START, payload: true });
  API.get(`/friends/${id}`)
    .then((res) => {
      dispatch({
        type: types.GET_USER_FRIENDS_SUCCESS,
        payload: res.data,
      });
    })
    .catch((err) => {
      dispatch({ type: types.GET_USER_FRIENDS_FAIL, payload: false });
    });
};

export const deleteUserFriends = (id, user) => (dispatch) => {
  dispatch({ type: types.DELETE_USER_FRIENDS_START, payload: true });
  API.delete(`friends/${id}`)
    .then((res) => {
      let newList = user.friends.filter((keep) => keep.id != id);
      dispatch({ type: types.DELETE_USER_FRIENDS_SUCCESS, payload: newList });
    })
    .catch((err) => {
      dispatch({ type: types.DELETE_USER_FRIENDS_FAIL, payload: false });
    });
};

export const addUserFriend = (user, friends_id) => (dispatch) => {
  dispatch({ type: types.ADD_USER_FRIEND_START, payload: true });
  API.post(`/friends`, { user_id: user.id, friends_id })
    .then((res) => {
      dispatch({ type: types.ADD_USER_FRIEND_SUCCESS, payload: false });
    })
    .then(() => {
      // dispatch `getUserFriends` to get updated list of friends
      dispatch(getUserFriends(user.id));
    })
    .catch((err) => {
      dispatch({
        type: types.ADD_USER_FRIEND_FAIL,
        payload: {
          isLoading: false,
          error: "There was an error adding your friend",
        },
      });
    });
};

// Promos
export const getUserPromos = (id) => (dispatch) => {
  dispatch({ type: types.GET_USER_FRIENDS_START, payload: true });
  API.get(`savedPromos/users/${id}`)
    .then((res) => {
      dispatch({
        type: types.GET_USER_PROMOS,
        payload: res.data,
      });
    })
    .catch((err) => {
      // Errors on actions need to be handeled some other way than
      // console.log(err);
    });
};

export const addUserPromo = (user_id, promo_id) => (dispatch) => {
  let postData = { user_id, promo_id };
  API.post("/savedPromos", postData)
    .then((res) => {
    })
    .catch((err) => {
      // Errors on actions need to be handeled some other way than
      // console.log(err);
    });
};

// Bio
export const updateUserBio = (changes) => (dispatch) => {
  API.put("/users", { bio: changes })
    .then((res) => {
      dispatch({ type: types.UPDATE_BIO_SUCCESS, payload: changes });
    })
    .catch((err) => {
      // Errors on actions need to be handeled some other way than
      // console.log(err.message);
    });
};
