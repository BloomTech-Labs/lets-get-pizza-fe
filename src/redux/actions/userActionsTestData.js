export const testData = {
  creds: {
    username: "JDawg",
    password: "1234",
  },
  message: "Welome pizzalover",
  token: "XXXXXXXXX",
  toggleClickSame: {
    event: { target: { id: "location" } },
    field: "location",
  },
  toggleClickDifferent: {
    event: { target: { id: "location" } },
    field: "dietary_preference",
  },
  settingsDietInput: {
    event: { target: {} },
    value: ["vegan"],
  },
  settingsAllInput: {
    event: { target: { name: "location" } },
    value: "L.A.",
  },
  user: {
    id: 1,
    bio: "I love my pizza",
    events: [
      {
        id: 1,
        user_id: 1,
        location_id: 1,
        title: "event1",
        description: "super amazing event",
        start_time: Date.now(),
        end_time: Date.now + 100,
      },
      {
        id: 2,
        user_id: 1,
        location_id: 1,
        title: "event2",
        description: "very terrible event",
        start_time: Date.now() + 1,
        end_time: Date.now + 101,
      },
    ],
    reviews: [
      {
        location_id: 80,
        user_id: 2,
        id: 52,
        rating: 5,
        review_title: "SOOOO GOOD",
        review_text: "yum yum yum!! GO HERE",
        business_name: "Rosa's Italian Resaurant",
        address: "2400 Columbus St, Bakersfield, CA 93306, United States",
      },
    ],
    friends: [
      {
        user_id: 1,
        friends_id: 1,
        status: "accepted",
        id: 1,
      },
      {
        user_id: 2,
        friends_id: 2,
        status: "accepted",
        id: 2,
      },
      {
        user_id: 3,
        friends_id: 3,
        status: "accepted",
        id: 3,
      },
    ],
    favShopDetails: {
      name: "Dominoes Pizza",
      latitude: 40.7050150708864,
      longitude: -73.9336165250072,
      address: "13 pizza rd",
      location_id: 25,
    },
    savedPromos: [
      {
        id: 1,
        location_id: 1,
        title: "Cheese Sale!",
        text: "Half Price extra cheese!",
        start_date: "2020-01-02T00:00:00.000Z",
        end_date: "2020-01-03T00:00:00.000Z",
      },
      {
        id: 2,
        location_id: 2,
        title: "Double Double Time!",
        text: "Buy one pizza and pay double for the second!",
        start_date: "2020-01-03T00:00:00.000Z",
        end_date: "2020-02-03T00:00:00.000Z",
      },
      {
        id: 3,
        location_id: 3,
        title: "Jailhouse Rock!",
        text: "Sing a song for 10 bucks off your order!",
        start_date: "2020-02-03T00:00:00.000Z",
        end_date: "2020-02-03T00:00:00.000Z",
      },
    ],
  },
  users: [
    {
      id: 1,
      username: "Billy",
      email: "billy@pizzaluv.com",
      password: "luckyGuess",
      profile_image:
        "https://res.cloudinary.com/plza/image/upload/v1588043869/qxhdqbj4sthf57bdgltz.jpg",
      display_name: "PizzaBruh420",
      dietary_preference: ["gluten"],
      favorite_pizza_toppings: "Mushrooms",
      display_location: "Miami, Florida",
      favorite_pizza_shop: 2,
      bio: "Pizza! Pizza! Pizza! Ain't nothing like it",
    },
    {
      id: 3,
      username: "Betty",
      email: "betty@pizzaluv.com",
      password: "luckyGuess",
      profile_image:
        "https://res.cloudinary.com/plza/image/upload/v1588043869/qxhdqbj4sthf57bdgltz.jpg",
      display_name: "PizzaGurl",
      dietary_preference: ["gluten"],
      favorite_pizza_toppings: "Pepperoni",
      display_location: "Billings",
      favorite_pizza_shop: 3,
      bio: "Can't spell pizza without Betty!",
    },
  ],
  save: { target: { id: "save" } },
  cancel: { target: { id: "cancel" } },
  locations: [
    {
      id: 1,
      last_name: "McTesterson",
      username: "DUNCAN",
      email: "test@test.com",
      password: "secret",
      first_name: "Jest",
      update_foursquare: true,
      phone_number: "5551234567",
      foursquare_id: "1",
      business_name: "Pizza King",
      latitude: 0,
      longitude: 0,
      address: "1234 Sauce St",
      website_url: "",
      official_description: "We like sauce like you like cheese",
      thumbnail_image: "",
      inside_image: "",
      street_view_image: "",
      menu_image: "",
      order_service: "horse & buggy",
      store_bio: "We reeeallly like sauce.",
      dietary_offerings: ["vegan"],
      events: [
        {
          description: "Its dat time boys, pizza!",
          end_time: "2020-01-22T00:00:00.000Z",
          id: 1,
          location_id: 1,
          start_time: "2020-01-20T00:00:00.000Z",
          title: "Lets Get Pizza!",
          user_id: 1,
        },
        {
          description: "Are Nacho pizzas a thing? Lets find out!",
          end_time: "2020-01-22T00:00:00.000Z",
          id: 2,
          location_id: 1,
          start_time: "2020-01-20T00:00:00.000Z",
          title: "Nacho Pizza Hunt!",
          user_id: 1,
        },
      ],
    },
  ],
  image: {
    id: 11,
    username: "pizzalover",
    email: "iluvpizza@test.com",
    profile_image:
      "https://res.cloudinary.com/plza/image/upload/v1592089926/bcdtzp5eoh9twubtrfmq.png",
    display_name: "PizzaLover",
    dietary_preference: null,
    favorite_pizza_toppings: "Pineapple",
    display_location: "Denver, CO",
    favorite_pizza_shop: 2,
    bio: "I could eat pizza for every meal",
  },
};
