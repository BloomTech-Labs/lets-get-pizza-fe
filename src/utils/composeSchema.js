import { object } from "yup";

/**
 * Compose Yup validation schema
 *
 * @param {object} initialSchema Example: `username`, `password`
 * @param {object} extraSchema
 * @returns {object} Composed object with initial and extra schema together
 */
export default function composeSchema(initialSchema, extraSchema) {
  return object().shape({ ...initialSchema, ...extraSchema });
}
