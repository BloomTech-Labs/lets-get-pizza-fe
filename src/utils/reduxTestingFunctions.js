// This function should take any component & an id
export const findElementById = (component, id) => {
  /*    FOR USE IN TESTING REDUX RENDERER   */
  // convert component to JSON so its easily managed with JavaScript
  const mother = component.toJSON();

  // Declare a variable to store current 'level' of looping
  // as we want to be able to loop as many times as the writer
  // has nested elements. The incoming component will always be a single
  // element, so its children must be nested in an array as if
  // we were accessing it from a parent.
  let currentLevel = [mother.children];

  // store the final item
  const filteredElement = [];

  // This function will dig into the arrays at the current level of nesting
  // and set the current level to that of its children.
  const repeatingLoop = () => {
    const updateLevel = [];
    currentLevel.forEach((arr) => {
      arr.forEach((object) => {
        if (object.children) {
          updateLevel.push(object.children);
        }
      });
    });
    currentLevel = updateLevel;
  };

  // This function will loop through all arrays of children at the current level
  // and will evaluate to see if any elements contain the given id from args.
  const lookForId = () => {
    currentLevel.forEach((arr) => {
      arr.forEach((object) => {
        if (object.props && object.props.id === id) {
          filteredElement.push(object);
        }
      });
    });
  };

  // This function will check to see if all children arrays are empty
  // in order to push an object with an error message so that the while
  // loop does not run forever if there is no id matching the id
  // recieved from args.
  const checkForStoppingPoint = () => {
    let combinedArr = 0;
    currentLevel.forEach((arr) => {
      if (arr.length !== 0) {
        combinedArr++;
      }
    });
    if (combinedArr === 0) {
      filteredElement.push({
        error: "no matching id found in the supplied component",
      });
    }
  };

  // Check to see if initial component object contains the desired id from args.
  if (mother.props.id === id) {
    return { ...mother.props, type: mother.type, children: mother.children };
  } else {
    // Otherwise run the functions to check if there are children,
    // if the current component contains the correct id, or if
    // the code needs to run at the next level of children.
    do {
      checkForStoppingPoint();
      lookForId();
      repeatingLoop();
      // do these functions until desired condition is met.
    } while (filteredElement.length === 0);
    return {
      ...filteredElement[0].props,
      type: filteredElement[0].type,
      children: filteredElement[0].children,
    };
  }
};

// When testing to see if a document contains text from props
// and I want to check that the data passed through correctly
// at least once, this function will find the first instance
export const findFirstProp = (textString, find, findAll) => {
  const regex = new RegExp(textString, "i");
  const array = findAll(regex);

  if (array.length <= 1) {
    return find(regex);
  } else {
    return array[0];
  }
};

/*    FOR USE IN TESTING REDUX ACTION CREATORS    */

// compares if items in an array are equal to each other
// requires an array of results & an array of expected
//      order of expectation array matters
// optional parameter of a single item not to be expected
export const compareExpectedCalls = (
  resultsArr,
  expectedArr,
  unexpectedItem
) => {
  resultsArr.forEach((result, idx) => {
    if (unexpectedItem) {
      expect(result).not.toEqual(unexpectedItem);
    }
    expect(result).toEqual(expectedArr[idx]);
  });
  expectedArr.forEach((expectation, idx) => {
    expect(expectation).toEqual(resultsArr[idx]);
  });
};

export const spreadCalls = (arrayOfArrays) => {
  let newArray = [];
  arrayOfArrays.forEach((array) => (newArray = [...newArray, ...array]));
    return newArray;
  };

export const compareExpectedState = (
  typeArray,
  initialState,
  payload,
  expectedState,
  reducer,
  failType
) => {
  // Map over given typeArray and return an array of action objects
  const actionsArray = typeArray.map((type) => ({ type, payload }));

  // Here we use .forEach() to run our test for each action in the array
  actionsArray.forEach((action) => {
    // Pass the initialState and action into the reducer
    // and set returnedState to the returned state from the reducer
    const returnedState = reducer(initialState, action);

    if (failType) {
      const failedState = reducer(initialState, { type: failType, payload });

      expect(failedState).not.toStrictEqual(expectedState);
    }

    expect(returnedState).toStrictEqual(expectedState);
    expect(returnedState).not.toStrictEqual(initialState);
    expect(returnedState).toBeDefined();
  });
};
