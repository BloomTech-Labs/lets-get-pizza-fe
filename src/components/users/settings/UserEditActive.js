import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { Dropdown, Icon, Input, Grid } from "semantic-ui-react";
import { userEditSettings } from "../../../redux/actions/userActions";
import UserEditButton from "./UserEditButton";

export default function UserEditActive({ item }) {
  const dispatch = useDispatch();
  const field = useSelector(({ user }) => user.field);

  const handleChange = (event, { value }) => {
    dispatch(userEditSettings(event, value));
  };

  return (
    <Grid.Row style={{ borderBottom: "1px solid lightgrey" }}>
      <Grid.Column style={{ display: "flex", alignItems: "center" }}>
        <Icon name={item.icon} size="large" />
        {field === "dietary_preference" ? (
          <Dropdown
            id="dropdown"
            multiple
            selection
            name="dietary_preference"
            placeholder="Dietary preference"
            options={[
              { text: "Vegan", value: "vegan" },
              { text: "Gluten-Free", value: "gluten-free" },
              { text: "Lacto-vegetarian", value: "lacto-vegetarian" },
              { text: "Ovo-vegetarian", value: "ovo-vegetarian" },
              { text: "Pescetarian", value: "pescetarian" },
              { text: "Vegetarian", value: "vegetarian" },
              { text: "None", value: "" },
            ]}
            onChange={handleChange}
          />
        ) : (
            <Input onChange={handleChange} name={item.name} value={item.value} />
          )}
      </Grid.Column>
      <UserEditButton text={"Save"} item={item} />
    </Grid.Row>
  );
}
