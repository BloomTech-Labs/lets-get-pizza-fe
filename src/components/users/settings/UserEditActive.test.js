import React from 'react';
import renderer from 'react-test-renderer';

import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { userEditSettings } from '../../../redux/actions/userActions';

import UserEditActive from './UserEditActive';
import { findElementById } from '../../../utils/reduxTestingFunctions';

jest.mock('../../../redux/actions/userActions');
const mockStore = configureStore();

describe('UserEditActive with item.value as a string', () => {

    let store;
    let component;

    const item = {
        icon: "location arrow",
        name: "location",
        value: "Barcelona"
    };
    const newValue = "Los Angeles"
    beforeEach(() => {
        store = mockStore({
            user: { field: 'location' }
        });

        store.dispatch = jest.fn();

        component = renderer.create(
            <Provider store={store}>
                <UserEditActive item={item} />
            </Provider>
        )

    });

    it('should render with text from props', () => {
        const inputElement = component.root.findByType('input').props

        expect(inputElement.name).toContain(item.name)
        expect(inputElement.value).toContain(item.value)
    })

    it('input onChange can fire userEditSettings for each option & fires correct Redux action', () => {
        renderer.act(() => {
            component.root.findByType('input').props.onChange({ target: { name: item.name } }, { value: newValue })
        });
        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith(userEditSettings({ target: { name: item.name } }, newValue));
    })

})

describe('UserEditActive with item.value as an array', () => {

    let store;
    let component;
    const item = {
        icon: "ban"
    };

    const options = [
        { text: "Vegan", value: "vegan" },
        { text: "Gluten-Free", value: "gluten-free" },
        { text: "Lacto-vegetarian", value: "lacto-vegetarian" },
        { text: "Ovo-vegetarian", value: "ovo-vegetarian" },
        { text: "Pescetarian", value: "pescetarian" },
        { text: "Vegetarian", value: "vegetarian" },
        { text: "None", value: "" },
    ]

    beforeEach(() => {
        store = mockStore({
            user: { field: 'dietary_preference' }
        });

        store.dispatch = jest.fn();

        component = renderer.create(
            <Provider store={store}>
                <UserEditActive item={item} />
            </Provider>
        )
    });

    it('should render with text from props', () => {
        const dropdown = findElementById(component, 'dropdown')
        expect(dropdown.name).toBe('dietary_preference')

        // find the dropdown items
        const menuItems = dropdown.children.filter(({ props }) => props.className && props.className === 'menu transition')[0]
        // loop through items & make sure they match incoming prop text
        menuItems.children.forEach((item, index) => {
            expect(item.children[0].children[0]).toContain(options[index].text)
        })
    })

    it('Dropdown onChange can fire userEditSettings for each option & fires correct Redux action', () => {

        renderer.act(() => {
            const dropdown = findElementById(component, 'dropdown')
            options.forEach(({ text, value }) => {
                dropdown.onChange({ target: { text: text } }, { value })
            })
        });

        expect(store.dispatch).toHaveBeenCalledTimes(options.length);
        expect(store.dispatch).toHaveBeenCalledWith(userEditSettings({ target: { text: 'Vegan' } }, 'vegan'));
    })
})