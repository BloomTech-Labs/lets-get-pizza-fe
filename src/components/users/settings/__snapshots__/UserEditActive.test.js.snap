// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserEditActive with item.value as a string should render with text from props 1`] = `
<div
  className="row"
  style={
    Object {
      "borderBottom": "1px solid lightgrey",
    }
  }
>
  <div
    className="column"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <i
      aria-hidden="true"
      className="location arrow large icon"
      onClick={[Function]}
    />
    <div
      className="ui input"
    >
      <input
        name="location"
        onChange={[Function]}
        type="text"
        value="Barcelona"
      />
    </div>
  </div>
  <div
    className="column"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <button
      className="ui tiny button"
      id="location"
      onClick={[Function]}
    >
      Save
    </button>
  </div>
</div>
`;

exports[`UserEditActive with item.value as an array should render with text from props 1`] = `
<div
  className="row"
  style={
    Object {
      "borderBottom": "1px solid lightgrey",
    }
  }
>
  <div
    className="column"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <i
      aria-hidden="true"
      className="ban large icon"
      onClick={[Function]}
    />
    <div
      aria-expanded={false}
      aria-multiselectable={true}
      className="ui multiple selection dropdown"
      id="dropdown"
      name="dietary_preference"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      role="listbox"
      tabIndex={0}
    >
      <div
        aria-atomic={true}
        aria-live="polite"
        className="default text"
        role="alert"
      >
        Dietary preference
      </div>
      <i
        aria-hidden="true"
        className="dropdown icon"
        onClick={[Function]}
      />
      <div
        className="menu transition"
      >
        <div
          aria-checked={false}
          aria-selected={true}
          className="selected item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            Vegan
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            Gluten-Free
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            Lacto-vegetarian
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            Ovo-vegetarian
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            Pescetarian
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            Vegetarian
          </span>
        </div>
        <div
          aria-checked={false}
          aria-selected={false}
          className="item"
          onClick={[Function]}
          role="option"
          style={
            Object {
              "pointerEvents": "all",
            }
          }
        >
          <span
            className="text"
          >
            None
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="column"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
      }
    }
  >
    <button
      className="ui tiny button"
      onClick={[Function]}
    >
      Save
    </button>
  </div>
</div>
`;
